src/cascadia/TerminalSettingsModel/FileUtils.cpp

```diff
@@ -125,7 +125,11 @@ namespace Microsoft::Terminal::Settings::Model
     {
         // GH#10787: In the case of symbolic link, ensure that we modify the target rather than the link itself.
         // We append the paths manually rather than using "canonical" method to support scenario in which link target doesn't exist
-        const auto resolvedPath = std::filesystem::is_symlink(path) ? path.parent_path() / std::filesystem::read_symlink(path) : path;
+        auto resolvedPath = path;
+        while (std::filesystem::is_symlink(resolvedPath))
+        {
+            resolvedPath = resolvedPath.parent_path() / std::filesystem::read_symlink(resolvedPath);
+        }

```

Comment:
One concern with manual symlink resolution is that it opens a can of worms regarding the symlink evaluation policy for L2L (local to local), L2R (local to remote), R2R, and R2L symlink types. (Check, for example, `fsutil.exe  behavior query symlinkEvaluation`.) I don't personally care about the first 3 types, but the last one is disturbing, and I think it should always be disabled. Symlinks are evaluated on the client side, so an R2L symlink is probably a mistake (e.g. "\\\\server\\share\\link" -> "C:\\local\\target\\path"). If a symlink type is disabled, then trying to follow it should fail with an `error_code` `value()` equal to `ERROR_SYMLINK_CLASS_DISABLED` (1463). This is the case for the `exists()` and `canonical()` methods, but `is_symlink()` and `read_symlink()` never ask the system to actually follow the link.

Reply:
@eryksun - these are great points: I didn't consider the remote links at all 🤦 But I am still not certain how to avoid manual resolution, if we want to support cases that the final target file doesn't exist, which I believe might be a common scenario (someone might remove the file to let the sessions be recreated)

Reply:
We could also try to resolve the file with `std::filesystem::canonical` and if the target doesn't exist (aka `canonical()` throws), fall back to writing the file directly, without the tmp file.
It allows atomic file operations for most users in most cases, but avoids the can of worms @eryksun has described.

Reply:
> 
> 
> We could also try to resolve the file with `std::filesystem::canonical` and if the target doesn't exist (aka `canonical()` throws), fall back to writing the file directly, without the tmp file.
> It allows atomic file operations for most users in most cases, but avoids the can of worms @eryksun has described.

My problem with this approach is the case when we have a link to a non-existing file.
In such case the canonical will fail and we will fallback to writing the file directly, rather than writing to the file that the link points to.
I believe that such scenario might be common: let's say I have a link to a settings folder (like in the original ticket) and I decided to delete the settings from this configuration folder, so they will be recreated. The suggested approach will destroy my file structure.

Reply:
I was thinking to use `canonical()`, and handle `ERROR_FILE_NOT_FOUND` (2), and only that error code, with manual symlink resolution. If the symlink type is disabled, the error will be `ERROR_SYMLINK_CLASS_DISABLED` (1463), which shouldn't be handled.

Reply:
Get it. I will do it. Thanks.

Reply:
> In such case the canonical will fail and we will fallback to writing the file directly, rather than writing to the file that the link points to.

I thought that if you `OpenFile()` a symlink on Windows it resolves that symlink. Do you mean to say that `OpenFile()` would overwrite the symlink? I thought that the rename operation is the problem here...

Reply:
@lhecker  - Oh.. Now I see what you mean. The question regarding the OpenFile and writing to it is whether the operation will be atomic enough (at least at the level this function tries to achieve).

Reply:
I mean, in your implementation:
```
        // GH#10787: In the case of symbolic link, ensure that we modify the target rather than the link itself.
        // If the link target does not exist, "canonical" will fail. In this case however we can simply write the file
        if (std::filesystem::exists(path))
        {
            const auto resolvedPath = std::filesystem::canonical(path);
            auto tmpPath = resolvedPath;
            tmpPath += L".tmp";

            // Writing to a file isn't atomic, but...
            WriteUTF8File(tmpPath, content);

            // renaming one is (supposed to be) atomic.
            // Wait... "supposed to be"!? Well it's technically not always atomic,
            // but it's pretty darn close to it, so... better than nothing.
            std::filesystem::rename(tmpPath, resolvedPath);
        }
        else
        {
            // Is this gooed enough? It is not atomic.
            WriteUTF8File(path, content);
        }
```

Reply:
> I thought that if you `OpenFile()` a symlink on Windows it resolves that symlink. Do you mean to say that `OpenFile()` would overwrite the symlink? I thought that the rename operation is the problem here...

By default, [`CreateFileW()`](https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-createfilew) follows all types of reparse points, unless it's passed the flag `FILE_FLAG_OPEN_REPARSE_POINT`. But the apparent goal here for the case of a symlink is to modify the symlink target in an atomic manner. So the new file has to be created beside the target, and then `rename()` is called to replace the target file.

Reply:
@eryksun, @lhecker  - what do you think about this:
```
        // GH#10787: In the case of symbolic link, ensure that we modify the target rather than the link itself.
        // If the link target does not exist, "canonical" will fail. In that case we resolve the link manually.
        auto resolvedPath = path;
        if (std::filesystem::exists(path))
        {
            resolvedPath = std::filesystem::canonical(path);
        }
        else
        {
            while (std::filesystem::is_symlink(resolvedPath))
            {
                resolvedPath = resolvedPath.parent_path() / std::filesystem::read_symlink(resolvedPath);
            }
        }

        auto tmpPath = resolvedPath;
        tmpPath += L".tmp";

        // Writing to a file isn't atomic, but...
        WriteUTF8File(tmpPath, content);

        // renaming one is (supposed to be) atomic.
        // Wait... "supposed to be"!? Well it's technically not always atomic,
        // but it's pretty darn close to it, so... better than nothing.
        std::filesystem::rename(tmpPath, resolvedPath);
```

Reply:
Personally I'd do:
```cpp
void WriteUTF8FileAtomic(const std::filesystem::path& path, const std::string_view content)
{
    // GH#10787: rename() will replace symlinks themselves and not the path they point at.
    // It's thus important that we first resolve them before generating tmpPath.
    std::error_code ec;
    const auto resolvedPath = std::filesystem::canonical(path, ec);

    if (ec)
    {
        // If path is a link and the target does not exist, "canonical" will fail.
        // Considering that our "atomic" writes are only on a best-effort base anyways,
        // the best fallback with highest compatibility is to just write the file.
        WriteUTF8File(path, content);
    }
    else
    {
        auto tmpPath = resolvedPath;
        tmpPath += L".tmp";

        // Writing to a file isn't atomic, but...
        WriteUTF8File(tmpPath, content);
        // renaming one is (supposed to be) atomic.
        // Wait... "supposed to be"!? Well it's technically not always atomic,
        // but it's pretty darn close to it, so... better than nothing.
        std::filesystem::rename(tmpPath, resolvedPath);
    }
}
```

Reply:
I'd manually resolve the path in the loop if `canonical(path)` raises a `filesystem_error` with a `code().value()` of `ERROR_FILE_NOT_FOUND`. Else `throw` the exception to let the caller handle the problem.

Reply:
@lhecker, @eryksun - you are both suggesting improved variations of two approaches attached earlier. Which is great, but we need to decide on one of them, and then we can improve it even further (e.g., by throwing if ec!=error_file_not_found). So the options are:
* @lhecker: if the file doesn't exist - create it
* @eryksun: if the file doesn't exist - resolve the path manually

The pro of @lhecker's approach is that path resolution will be more accurate. The pro of @eryksun's approach is that we preserve better atomicity when the file doesn't exist. What I am missing is how much this atomicity important when the file doesn't exist.

Reply:
> What I am missing is how much this atomicity important when the file doesn't exist.

The intent is to avoid having an accessible file that's in the process of being written, so one uses a temporary name until the file is completely written. (Usually the name is a relatively long random name, rather than just appending ".tmp" to the  target name.) The alternative in Windows is to to open without sharing any data access (not generally possible in POSIX), in which case other threads that try to open the file will fail with `ERROR_SHARING_VIOLATION` (32), but that's more complicated for everyone.

Reply:
@eryksun - I understand a concept of atomic update and the use of renaming as a way to simulate it. 
My question was: what is the impact of this function not being atomic in the case that the file doesn't exist. E..g.: does it mean that every 3rd fresh run will fail, or does it mean that hypothetically in some weird case there will be some message.

Reply:
Ever since @eryksun brought up the topic of symlink evaluation policies I'm convinced that we should avoid manually resolving symlinks. Because if the resolve-loop is fine as a fallback I don't see why we should use `canonical()` in the first place - the "vulnerability" would still be there as a fallback.
In either case, an imperfect solution is fine for now in my opinion. If possible I'd err on the side of security wherever we can though.

Reply:
I'm only looking at from the POV of implementing the intent of the function. The implementation of `rename()` in Windows calls `MoveFileExW()` with the flags `MOVEFILE_COPY_ALLOWED` and `MOVEFILE_REPLACE_EXISTING`. It won't be atomic if it has to copy the file across filesystems and delete the original, but renaming a file within a filesystem is atomic.

Reply:
> Because if the resolve-loop is fine as a fallback I don't see why we should use `canonical()` in the first place - the "vulnerability" would still be there as a fallback.

The only error I'd handle from `canonical()` is a missing target file, in which case the symlink type was supported -- at the time of the `canonical()` call. However, there is some concern that another thread can meddle with the symlink(s) in between calls (e.g. replace a link with one that targets another path) while resolving the path manually.

Reply:
@eryksun - I am not sure we need to preserve the intent of the function at any price, as it is used in 2 places only in the code. I also agree with @lhecker that manual resolution of the symlinks is risky. My major concern with @lhecker's approach is when will this lack of atomicity screw us? For instance we have a watchdog on the folder - will it get triggered in the middle of the writing the new file, will it crash the terminal?

Reply:
@eryksun, @lhecker - I pushed the version that:
1. Uses canonical by default
2. Throws if canonical fails not due to file not found
3. Has a single clause that handles the corner case when the path points to a link and the target doesn't exist
4. I have chosen @lhecker's suggestion of implementing this fallback clause by simply writing to a file